File: packages/android-client/FishjamClient/src/main/java/com/fishjamcloud/client/webrtc/RTCEngineListener.kt
package com.fishjamcloud.client.webrtc

import com.fishjamcloud.client.models.Metadata
import fishjam.media_events.Shared
import fishjam.media_events.server.Server

internal interface RTCEngineListener {
  fun onConnected(
    endpointId: String,
    endpoints: Map<String, Server.MediaEvent.Endpoint>,
    iceServers: List<Server.MediaEvent.IceServer>
  )

  fun onSendMediaEvent(event: fishjam.media_events.peer.Peer.MediaEvent)

  fun onEndpointAdded(
    endpointId: String,
    metadata: Metadata?
  )

  fun onEndpointRemoved(endpointId: String)

  fun onEndpointUpdated(
    endpointId: String,
    endpointMetadata: Metadata? = mapOf()
  )

  fun onOfferData(tracksTypes: Server.MediaEvent.OfferData.TrackTypes)

  fun onSdpAnswer(
    sdp: String,
    midToTrackId: Map<String, String>
  )

  fun onRemoteCandidate(
    candidate: String,
    sdpMLineIndex: Int,
    sdpMid: String
  )

  fun onTracksAdded(
    endpointId: String,
    trackIdToTrack: Map<String, Server.MediaEvent.Track>
  )

  fun onTracksRemoved(
    endpointId: String,
    trackIds: List<String>
  )

  fun onTrackUpdated(
    endpointId: String,
    trackId: String,
    metadata: Metadata? = mapOf()
  )

  fun onVadNotification(
    trackId: String,
    status: Server.MediaEvent.VadNotification.Status
  )

  fun onBandwidthEstimation(estimation: Long)
}

File: packages/ios-client/Sources/FishjamClient/webrtc/RTCEngineListener.swift
internal protocol RTCEngineListener: AnyObject {
    func onSendMediaEvent(event: Fishjam_MediaEvents_Peer_MediaEvent.OneOf_Content)
    func onConnected(
        endpointId: String, endpointIdToEndpoint: [String: Fishjam_MediaEvents_Server_MediaEvent.Endpoint],
        iceServers: [Fishjam_MediaEvents_Server_MediaEvent.IceServer])
    func onEndpointAdded(endpointId: String, metadata: Metadata?)
    func onEndpointRemoved(endpointId: String)
    func onEndpointUpdated(endpointId: String, metadata: Metadata?)
    func onOfferData(tracksTypes: Fishjam_MediaEvents_Server_MediaEvent.OfferData.TrackTypes)
    func onSdpAnswer(sdp: String, midToTrackId: [String: String])
    func onRemoteCandidate(candidate: String, sdpMLineIndex: Int32, sdpMid: String?)
    func onTracksAdded(endpointId: String, trackIdToTracks: [String: Fishjam_MediaEvents_Server_MediaEvent.Track])
    func onTracksRemoved(endpointId: String, trackIds: [String])
    func onTrackUpdated(endpointId: String, trackId: String, metadata: Metadata)
    func onTrackEncodingChanged(endpointId: String, trackId: String, encoding: String, encodingReason: String)
    func onVadNotification(trackId: String, status: Fishjam_MediaEvents_Server_MediaEvent.VadNotification.Status)
    func onBandwidthEstimation(estimation: Int)
}
